src/Microsoft.DotNet.DeadRegionAnalysis/AnalysisEngine.RegionRemoval.cs

```diff
@@ -0,0 +1,196 @@
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.DotNet.DeadRegionAnalysis
+{
+    public partial class AnalysisEngine
+    {
+        public async Task<Document> RemoveUnnecessaryRegions(DocumentConditionalRegionInfo info, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (info == null)
+            {
+                throw new ArgumentNullException("info");
+            }
+
+            var changes = CalculateTextChanges(info.Chains);
+            if (changes == null || changes.Count == 0)
+            {
+                return info.Document;
+            }
+
+            var newText = await info.Document.GetTextAsync(cancellationToken);
+
+            try
+            {
+                newText = newText.WithChanges(changes);
+            }
+            catch (Exception)
+            {
+                var changesString = new StringBuilder();
+                var syntaxTree = info.Document.GetSyntaxTreeAsync(cancellationToken).Result;
+
+                foreach (var change in changes)
+                {
+                    var lineSpan = Location.Create(syntaxTree, change.Span).GetLineSpan();
+                    changesString.AppendFormat("({0}-{1}): {2}", lineSpan.StartLinePosition.Line, lineSpan.EndLinePosition.Line, newText.GetSubText(change.Span).ToString());
+                }
+
+                Console.WriteLine(string.Format("Failed to remove regions from document '{0}':{1}{2}", info.Document.FilePath, Environment.NewLine, changesString.ToString()));
+                return info.Document;
+            }
+
+            return info.Document.WithText(newText);
+        }
+
+        private static List<TextChange> CalculateTextChanges(List<ConditionalRegionChain> chains)
+        {
+            var changes = new List<TextChange>();
+
+            foreach (var chain in chains)
+            {
+                CalculateTextChanges(chain, changes);
+            }
+
+            changes.Sort(CompareTextChanges);
+            return MergeOverlappingRegions(changes);
+        }
+
+        public static void CalculateTextChanges(ConditionalRegionChain chain, List<TextChange> changes)
+        {
+            bool removeEndif = true;
+
+            for (int i = 0; i < chain.Regions.Count; i++)
+            {
+                var region = chain.Regions[i];
+                if (region.State != Tristate.Varying)
+                {
+                    var startDirective = region.StartDirective;
+                    var endDirective = region.EndDirective;
+                    string endDirectiveReplacementText = string.Empty;
+
+                    // Remove the start directive
+                    changes.Add(new TextChange(new TextSpan(region.SpanStart, region.StartDirective.FullSpan.End - region.SpanStart), string.Empty));
```

Comment:
region.StartDirective.FullSpan does not include any whitespace trivia preceding the DirectiveTriviaSyntax.  That's the reason for the region.SpanStart property.